### PR TITLE
parser: fix required function members in struct

### LINKED ISSUE #10290: Cannot declare a function as a required struct member
**V doctor:** 
```
OS: linux, "Artix Linux"
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 3600 6-Core Processor
CC version: cc (GCC) 11.1.0

getwd: /home/me/coding/v/nice_server
vmodules: /home/me/.vmodules
vroot: /home/me/Programs/v
vexe: /home/me/Programs/v/v
vexe mtime: 2021-05-31 15:13:35
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 e4f6369.90292ce

Git version: git version 2.31.1
Git vroot status: weekly.2021.21-74-g90292ce9
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v run test.v`
```

struct Struct{
	a int [required]
	b fn() [required]
}

fn main(){
	s := Struct{}
}
```

**What did you expect to see?**
No errors
 
**What did you see instead?**
```
test.v:4:10: error: non-constant array bound `required`
    2 | struct Struct{
    3 |     a int [required]
    4 |     b fn() [required]
      |             ~~~~~~~~
    5 | }
    6 |
```

### PR DESCRIPTION
**Pull request by eafzali**
fixes #10290

I first implemented a simpler way in https://github.com/vlang/v/pull/10299/commits/20b1fdd2538fec57b0fa5b2dca15fb8a1830b33c but realized that it's not covering all edge cases like 

```v
struct Struct {
 	f fn () [custom_attr; required]
}
```
or
```v
const required = 12

struct Struct {
	f fn () [required]int
}
```

Also filed #10320 for the above example.

I've also considered to use something like `(p.tok.kind != .lsbr || p.peek_tok.kind == .rsbr)` in https://github.com/vlang/v/pull/10299#discussion_r643130504 but for and example like:

```v
struct Struct {
 	f fn () [12]int
 }

 fn main() {
 	s := Struct{
 	}
	
	println(s)
 }
```
I'm getting
```
m.v:2:12: error: unexpected number `12`, expecting name
    1 | struct Struct {
    2 |      f fn () [12]int
      |               ~~
    3 |  }
    4 |
```
Which is very confusing and not user friendly so, I moved forward with implementing is_attributes.

### CODE REVIEW

File: vlib/v/parser/parse_type.v [hunk 1]
```diff
@@ -212,7 +212,10 @@ pub fn (mut p Parser) parse_fn_type(name string) ast.Type {
 	}
 	mut return_type := ast.void_type
 	mut return_type_pos := token.Position{}
-	if p.tok.line_nr == line_nr && p.tok.kind.is_start_of_type() {
+
+	// don't confuse required attribute: struct{ f fn () [required] }
+	if p.tok.line_nr == line_nr && p.tok.kind.is_start_of_type()
+		&& (p.tok.kind != .lsbr || p.peek_tok.lit != 'required') {
```
Comment (eafzali) [inline vlib/v/parser/parse_type.v:218]: If functions are not allowed to return fixed sized arrays:

```suggestion
	// return type can not be a fixed size array, but there might be an
	// attribute list after fn_type in struct_decl: struct{ f fn () [required] }
	if p.tok.line_nr == line_nr && p.tok.kind.is_start_of_type()
		&& (p.tok.kind != .lsbr || p.peek_tok.kind == .rsbr) {
```

**Comment by spytheman** [comment]
You can have arbitrary lookahead by calling `p.peek_token(n)` . `p.peek_tok` is just a cached version of `p.peek_token(0)`, because it is the most commonly used one.

File: vlib/v/parser/parse_type.v (hunk 1 above)
Comment (eafzali) [reply vlib/v/parser/parse_type.v:218]: @spytheman   can a function return fixed size array?

Do you think the above suggestion works?

If not, does it make sense to add something like 

```v
fn (mut p Parser) is_attributes() bool {
	if p.tok.kind != .lsbr {
		return false
	}

	mut i := 0
	for p.peek_token(i).kind != .rsbr {
		i++
	}
	peek_rsbr_tok:= p.peek_token(i+1)
	if peek_rsbr_tok.line_nr == p.tok.line_nr && peek_rsbr_tok.kind != .rcbr {
		return false
	}

	return true
}
```

in `parser.v` and then apply this suggestion:

```suggestion
	if p.tok.line_nr == line_nr && p.tok.kind.is_start_of_type() && !p.is_attributes() {
```

Can you suggest any better approach?

Comment (JalonSolov) [reply vlib/v/parser/parse_type.v:218]: No, you cannot return fixed arrays, because they are allocated on the stack.  Return a clone of it, instead.

Comment (eafzali) [reply vlib/v/parser/parse_type.v:218]: Thanks for confirming @JalonSolov. I don't want to return a fixed array, I was asking to make sure if that's the case, then I could do https://github.com/vlang/v/pull/10299#discussion_r643130504 to not confuse attributes with array in fn return in struct declaration

**Comment by eafzali** [comment]
added #10320 for fixed sized array return in functions

File: vlib/v/parser/parse_type.v (hunk 1 above)
Comment (eafzali) [reply vlib/v/parser/parse_type.v:218]: If I use this suggestion to use `(p.tok.kind != .lsbr || p.peek_tok.kind == .rsbr)`, for this:

```v
struct Struct {
 	f fn () [12]int
 }

 fn main() {
 	s := Struct{
 	}
	
	println(s)
 }
```

I'm getting 
```
m.v:2:12: error: unexpected number `12`, expecting name
    1 | struct Struct {
    2 |      f fn () [12]int
      |               ~~
    3 |  }
    4 |
```

Which is very confusing and not user friendly so, I'm giving up on this approach and moving forward with `is_attributes`.

**Comment by medvednikov** [comment]
The syntax is

```
field type attribute
```

so the correct syntax for this example would be
```
struct Struct {
	f fn () int [required]
}
```

File: vlib/v/tests/struct_fields_storing_required_functions_test.v [hunk 2]
```diff
@@ -0,0 +1,17 @@
+struct Struct {
+	f1 fn () [required]
+	f2 fn () [attr1; required]
```
Comment (eafzali) [inline vlib/v/tests/struct_fields_storing_required_functions_test.v:3]: @medvednikov I assume this is also accepted as a correct syntax, right? Can developers use custom attributes and use reflection to get them? or maybe introducing more builtin supported attributes for struct fields?

https://github.com/vlang/v/pull/10299/commits/20b1fdd2538fec57b0fa5b2dca15fb8a1830b33c was much simpler but I don't think was covering all edge cases.

Comment (medvednikov) [reply vlib/v/tests/struct_fields_storing_required_functions_test.v:3]: Yes this is the syntax I described in my last message.

**Review by spytheman — APPROVED** [review]
Good work.
